services/circleci/circleci.helpers.js

```diff
@@ -0,0 +1,121 @@
+'use strict'
+
+const Joi = require('joi')
+
+const getWorkflowId = function(build) {
+  return build.workflows.workflow_id
+}
+
+// find all the builds with the same workflow id
+const getBuildsForWorkflow = function(builds, workflowId) {
+  return builds.filter(build => getWorkflowId(build) === workflowId)
+}
+
+// find all the builds with the same workflow id which are also complete
+const getCompleteBuildsForWorkflow = function(builds, workflowId) {
+  return builds.filter(
+    build => getWorkflowId(build) === workflowId && build.outcome !== null
+  )
+}
+
+// Find the most recent workflow which contains only complete builds
+// and return all the builds with that workflow id
+const getBuildsForLatestCompleteWorkflow = function(builds) {
+  let allBuilds, completeBuilds
+  for (let i = 0; i < builds.length; i++) {
+    allBuilds = getBuildsForWorkflow(builds, getWorkflowId(builds[i]))
+    completeBuilds = getCompleteBuildsForWorkflow(
+      builds,
+      getWorkflowId(builds[i])
+    )
+    if (allBuilds.length === completeBuilds.length) {
+      return completeBuilds
+    }
+  }
+  throw new Error('No complete workflows found')
+}
+
+const countOutcomes = function(builds) {
+  let total = 0
+  const counts = {
+    canceled: 0,
+    infrastructure_fail: 0,
+    timedout: 0,
+    failed: 0,
+    no_tests: 0,
+    success: 0,
+  }
+  for (let i = 0; i < builds.length; i++) {
+    if (!(builds[i].outcome in counts)) {
+      throw new Error('Found unexpected outcome')
+    }
+    counts[builds[i].outcome]++
+    total++
+  }
+  return { total, counts }
+}
+
+// reduce the outcomes for an array of builds to a single status
+const summarizeBuilds = function(builds) {
+  const { total, counts } = countOutcomes(builds)
+
+  if (total === counts.success) {
+    return 'passing'
+  } else if (counts.no_tests >= 1) {
+    return 'no tests'
+  } else if (counts.infrastructure_fail >= 1) {
+    return 'infrastructure fail'
+  } else if (counts.canceled >= 1) {
+    return 'canceled'
+  } else if (counts.timedout >= 1) {
+    return 'timed out'
+  } else if (counts.failed >= 1) {
+    return 'failed'
```

Comment:
I've taken a stab at a sensible order here, but I'm not committed to it. If you're reading this thinking "'canceled' obviously takes precedence over 'infrastructure fail'", feel free to make suggestions and I'll change it.

I've also switched to using 'outcome' here instead of status (outcome only applies to finished builds and we only want to summarise workflows where all the builds are finished). This also gives as smaller number of values to try and make sense of.